test/powershell/Modules/Microsoft.PowerShell.Utility/XMLCommand.Tests.ps1

```diff
@@ -340,4 +346,181 @@ Describe "XmlCommand DRT basic functionality Tests" -Tags "CI" {
+		$deserialized_one = ConvertFrom-CliXml -InputObject $content
+		$deserialized_one.value | Should -Be 1
+        $deserialized_one.two | Should -Not -BeNullOrEmpty
+		$deserialized_one.two.value | Should -Be 2
+		$deserialized_one.two.three | Should -Not -BeNullOrEmpty
+		$deserialized_one.two.three.num | Should -BeNullOrEmpty
+    }
+
+    It "ConvertTo-CliXml using -Depth 3 should work" {
+        $testObject = [One]::New()
+        $content = $testObject | ConvertTo-CliXml -Depth 3
+        $testObject | Export-CliXml -Path $testfile -Depth 3
+        $content | Should -Be (Get-Content -Path $testfile -Raw)
+        $deserialized_one = ConvertFrom-CliXml -InputObject $content
+        $deserialized_one.value | Should -Be 1
+        $deserialized_one.two | Should -Not -BeNullOrEmpty
+        $deserialized_one.two.value | Should -Be 2
+        $deserialized_one.two.three | Should -Not -BeNullOrEmpty
+        $deserialized_one.two.three.num | Should -Be 3
+        $deserialized_one.two.three.four | Should -Not -BeNullOrEmpty
+        $deserialized_one.two.three.four.num | Should -BeNullOrEmpty
+    }
+
+    It "ConvertTo-CliXml should serialize array correctly using ValueFromPipeline" {
+        $testObject = @(1,2,3,4)
+        $content = $testObject | ConvertTo-CliXml
+        $testObject | Export-CliXml -Path $testfile
+        $content | Should -Be (Get-Content -Path $testfile -Raw)
+        $out = ConvertFrom-CliXml -InputObject $content
+        $out -is [array] | Should -BeTrue
+        $out.Count | Should -Be 4
+        $out[0] | Should -Be 1
+        $out[1] | Should -Be 2
+        $out[2] | Should -Be 3
+        $out[3] | Should -Be 4
+    }
+
+    It "ConvertTo-CliXml should serialize array correctly using -InputObject" {
+        $testObject = @(1,2,3,4)
+        $content = ConvertTo-CliXml -InputObject $testObject
+        Export-CliXml -Path $testfile -InputObject $testObject
+        $content | Should -Be (Get-Content -Path $testfile -Raw)
+        $out = ConvertFrom-CliXml -InputObject $content
+        $out -is [System.Collections.ArrayList] | Should -BeTrue
+        $out.Count | Should -Be 4
+        $out[0] | Should -Be 1
+        $out[1] | Should -Be 2
+        $out[2] | Should -Be 3
+        $out[3] | Should -Be 4
+    }
+
+    It "ConvertTo-CliXml should serialize hashtable correctly" {
+        $testObject = [ordered]@{ a = 1; b = 2; c = 3; d = 4 }
+        $content = $testObject | ConvertTo-CliXml
+        $testObject | Export-CliXml -Path $testfile
+        $content | Should -Be (Get-Content -Path $testfile -Raw)
+        $out = ConvertFrom-CliXml -InputObject $content
+        $out -is [ordered] | Should -BeTrue
+        $out.Count | Should -Be 4
+        $out.Keys | Should -BeIn @('a', 'b', 'c', 'd')
+        $out.Values | Should -BeIn @(1, 2, 3, 4)
+    }
+
+    It "ConvertTo-CliXml should serialize PSCustomObject correctly" {
+        $testObject = [PSCustomObject]@{ a = 1; b = 2; c = 3; d = 4 }
```

Comment:
Maybe use one nested object?

Reply:
Added test for nested object in d90b55133